### PR TITLE
Feature/handle situation of single charge

### PR DESCRIPTION
**Pull request by CodeBear801**
# Issue
https://github.com/Telenav/osrm-backend/issues/143

## Tasklist

 - [x] Implement logic
 - [x] add tests 
 - [ ] review
 - [x] adjust for comments

## Requirements / Relations

Please go to https://github.com/Telenav/osrm-backend/issues/141

### CODE REVIEW

File: integration/oasis/handler.go [hunk 1]
```diff
@@ -45,6 +48,15 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
+	// check whether orig and dest is reachable
+	if len(route.Routes) == 0 {
+		glog.Info("Orig and destination is not reachable for request.")
```
Comment (wangyoucao577) [inline integration/oasis/handler.go]: Nit: Better to provide more information here, i.e. provide both origin and destination coordinates. The idea is from https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures but I think it's also applicabe for any logs.

File: integration/oasis/handler.go [hunk 2]
```diff
@@ -55,9 +67,18 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
 		h.generateOASISResponse(w, route, remainRange)
+		return
+	}
+
+	// check whether could achieve by single charge
+	overlap := getOverlapChargeStations4OrigDest(oasisReq, route.Routes[0].Distance, h.osrmConnector, h.tnSearchConnector)
+	if len(overlap) > 0 {
+		generateResponse4SingleChargeStation(w, oasisReq, overlap, h.osrmConnector)
+		return
 	} else {
```
Comment (wangyoucao577) [inline integration/oasis/handler.go]: `else` seems useless here. See https://golang.org/doc/effective_go.html#if

File: integration/oasis/reachable_by_single_charge.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	"github.com/Telenav/osrm-backend/integration/oasis/searchconnector"
+	"github.com/Telenav/osrm-backend/integration/oasis/stationfinder"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/oasis"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/osrm/coordinate"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/osrm/table"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/search/nearbychargestation"
+	"github.com/golang/glog"
+)
+
+const maxOverlapPointsNum int = 500
```
Comment (wangyoucao577) [inline integration/oasis/reachable_by_single_charge.go]: Nit: `int` is unncessary here for `const`.

File: integration/oasis/stationfinder/basic_finder.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package stationfinder
+
+import (
+	"sync"
+
+	"github.com/Telenav/osrm-backend/integration/pkg/api/search/nearbychargestation"
+)
+
+type basicFinder struct {
```
Comment (wangyoucao577) [inline integration/oasis/stationfinder/basic_finder.go]: why empty struct?

**Comment by wangyoucao577** [comment]
By the way, the "Nit:" comes from https://github.com/google/eng-practices/blob/master/review/reviewer/looking-for.md#style and I think it's a good way to say suggestion only, not madantory to change.

**Comment by CodeBear801** [comment]
Information from https://github.com/google/eng-practices/blob/master/review/reviewer/looking-for.md is very helpful

- The most important thing to cover in a review is the overall design of the CL
- Be careful of what's the impact to final user: logic, UI, flow etc
- Be careful of parallel programming going on in the CL that could theoretically cause deadlocks or race conditions.  Need people to think though
- Be aware of complexity
- comments are useful when they explain why some code exists, and should not be explaining what some code is doing.

**Pull request merged**